subprojects/core-api/src/main/java/org/gradle/api/IsolatedProject.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+
+/**
+ * An isolated view of {@link Project} that exposes only those properties that are safe to access from outside of
+ * <code>this</code> project, from the perspective of isolated projects.
+ * @since 8.7
+ */
+@Incubating
+public interface IsolatedProject {
+
+    /**
+     * <p>Returns the name of this project. The project's name is not necessarily unique within a project hierarchy. You
+     * should use the {@link #getPath()} method for a unique identifier for the project.
+     * If the root project is unnamed and is located on a file system root it will have a randomly-generated name
+     * </p>
+     *
+     * @return The name of this project. Never return null.
+     * @since 8.7
+     */
+    @NotNull
+    String getName();
+
+    /**
+     * <p>Returns the path of this project.  The path is the fully qualified name of the project.</p>
+     *
+     * @return The path. Never returns null.
+     * @since 8.7
+     */
+    @NotNull
+    String getPath();
+
+    /**
+     * <p>The directory containing the project build file.</p>
+     *
+     * @return The project directory. Never returns null.
+     * @since 8.7
+     */
+    @NotNull
+    File getProjectDirectory();
+
+    /**
+     * <p>Returns the root directory of this project. The root directory is the project directory of the root
+     * project.</p>
+     *
+     * @return The root directory. Never returns null.
+     * @since 8.7
+     */
+    @NotNull
+    File getRootDirectory();
```

Comment:
This seems like a weird property. 
Seems like we should have a `getParent` instead of this

Reply:
Indeed. It's going to be removed.

Reply:
The original intent was to provide a canonical API that was gradle-blessed to be safe to access from the isolated projects perspective. It grew out of an internal API I wrote at my company based on conversations that said it was ok to access the root directory from a project. This is far from obvious to end users, since it at least looks like a File in another project's hierarchy. So, if this gets removed, I'd at least love to see something on the javadoc of `Project.getRootDir()` that tells users it's ok to use.

Reply:
> Indeed. It's going to be removed.

Reply:
> (maybe) a getRootProject for convenience

For safe access to root project dir with isolated projects would that mean using `project.isolated.rootProject.isolated.projectDirectory` to ensure safety if just relying on the API shape? Not using getIsolated would still mean violations would be reported when Gradle is invoked but having convenient methods in the isolated API makes it easier to see at a glance that things are done properly.

Or would it be `rootProject.isolated.projectDirectory`?

Javadoc updates to say which methods are safe to call using IP is an alternative but less discoverable - but should be done anyway so things like creating Gradle plugins that support earlier versions (without IsolatedProject support) is easier.

Reply:
`project.isolated.rootProject.projectDirectory` is now equivalent to `project.rootProject.isolated.projectDirectory`